# GoGreen increases x% participation in future Marketing campaigns with Machine Learning

## Executive summary


## Background

GoGreen is a company in the solar industry.  It has hired us as consultants to help them understand which factors contribute to the success of future marketing campaigns.  Currently the company is trying to reduce marketing efforts that result in zero return on investment (ROI).
We are also tasked to answer the following questions: What does the Average customer look like for our company? What Products and Channels of Revenue are best performing? Which Marketing Campaigns were most successful? 

## Overview of Approach

We will answer these questions through data visualization, statistical analysis, and several machine learning models leveraging Python.

### Data Dictionary

The following are the fields that we have in our data:


- `ID`: the unique identification code for every customer
- `Year_Birth`: The Year of a customer's birth
- `Education`: The level of education that a customer completed
- `Marital_Status`: Status of Marriage
- `Income`: Annual Income
- `Kidhome`: # of children under the age of 13 in Customer's household
- `Teenhome`: # of children between 13-19 in Customer's household
- `Dt_Customer`: Date of Customer Enrollment
- `Recency`: # of days since last purchase
- `MntWines`: Dollar amount of Wines purchased in last 2 years
- `MntFruits`: Dollar amount of Fruits purchased in last 2 years
- `MntMeatProducts`: Dollar amount of Meat products purchased in the last 2 years
- `MntFishProducts`: Dollar amount of Fish products purchased in the last 2 years
- `MntSweetProducts`: Dollar amount of Sweet products purchased in the last 2 years
- `MntGoldProds`: Dollar amount of Gold products purchased in the last 2 years
- `NumDealsPurchases`: # of purchases made with discount
- `NumWebPurchases`: # of purchases made through the company's website
- `NumCatalogPurchases`: # of purchases made using the catalog
- `NumStorePurchases`: # of purchases made directly in-store
- `NumWebVisitsMonth`: # of visits made through company's website
- `AcceptedCmp1`: 1 if customer accepted the offer in the 1st campaign, 0 otherwise
- `AcceptedCmp2`: 1 if customer accepted the offer in the 2nd campaign, 0 otherwise
- `AcceptedCmp3`: 1 if customer accepted the offer in the 3rd campaign, 0 otherwise
- `AcceptedCmp4`: 1 if customer accepted the offer in the 4th campaign, 0 otherwise
- `AcceptedCmp5`: 1 if customer accepted the offer in the 5th campaign, 0 otherwise
- `Complain`: 1 if customer complained in the last 2 years, 0 otherwise
- `Response`: 1 if customer accepted the offer in the last campaign, 0 otherwise


## Let's start coding!

### Import Libraries

In [1]:
# Libraries to help with reading and manipulating data
import numpy as np
import pandas as pd

In [2]:
# Libraries to help with data visualization
import matplotlib.pyplot as plt
import seaborn as sns

sns.set()

In [3]:
# Removes the limit for the number of displayed columns
pd.set_option("display.max_columns", None)
# Sets the limit for the number of displayed rows
pd.set_option("display.max_rows", 200)

In [4]:
# To supress numerical display in scientific notations
pd.set_option("display.float_format", lambda x: "%.2f" % x)

In [5]:
# Avoid displaying warnings
import warnings

In [6]:
# Machine Learning Libraries



In [7]:
# this will help in making the Python code more structured automatically (good coding practice)
%load_ext nb_black

<IPython.core.display.Javascript object>

In [8]:
# loading the dataset
df = pd.read_excel("marketing_campaign.xlsx")

<IPython.core.display.Javascript object>

In [9]:
# checking shape of the data
print(f"There are {df.shape[0]} rows and {df.shape[1]} columns in the dataset")

There are 2240 rows and 29 columns in the dataset


<IPython.core.display.Javascript object>

In [10]:
df.head()

ID  Year_Birth   Education Marital_Status   Income  Kidhome  Teenhome  \
0  5524        1957  Graduation         Single 58138.00        0         0   
1  2174        1954  Graduation         Single 46344.00        1         1   
2  4141        1965  Graduation       Together 71613.00        0         0   
3  6182        1984  Graduation       Together 26646.00        1         0   
4  5324        1981         PhD        Married 58293.00        1         0   

  Dt_Customer  Recency  MntWines  MntFruits  MntMeatProducts  MntFishProducts  \
0  2012-09-04       58       635         88              546              172   
1  2014-03-08       38        11          1                6                2   
2  2013-08-21       26       426         49              127              111   
3  2014-02-10       26        11          4               20               10   
4  2014-01-19       94       173         43              118               46   

   MntSweetProducts  MntGoldProds  NumDealsPurchases  NumWebPurchases  \
0                88            88                  3                8   
1                 1             6                  2                1   
2                21            42                  1                8   
3                 3             5                  2                2   
4                27            15                  5                5   

   NumCatalogPurchases  NumStorePurchases  NumWebVisitsMonth  AcceptedCmp3  \
0                   10                  4                  7             0   
1                    1                  2                  5             0   
2                    2                 10                  4             0   
3                    0                  4                  6             0   
4                    3                  6                  5             0   

   AcceptedCmp4  AcceptedCmp5  AcceptedCmp1  AcceptedCmp2  Complain  \
0             0             0             0             0         0   
1             0             0             0             0         0   
2             0             0             0             0         0   
3             0             0             0             0         0   
4             0             0             0             0         0   

   Z_CostContact  Z_Revenue  Response  
0              3         11         1  
1              3         11         0  
2              3         11         0  
3              3         11         0  
4              3         11         0

<IPython.core.display.Javascript object>

In [32]:
# to view last 5 rows of the dataset
df.tail(100)

ID  Year_Birth   Education Marital_Status    Income  Kidhome  \
2140  10432        1974  Graduation       Divorced  19346.00        1   
2141   9216        1971  Graduation        Married  35788.00        1   
2142   7124        1968  Graduation       Divorced  36997.00        1   
2143   9727        1957  Graduation        Married  23539.00        0   
2144   5136        1973  Graduation         Single  65333.00        0   
2145   9790        1957  Graduation         Single  78499.00        0   
2146   1818        1971         PhD       Together  29732.00        1   
2147   1100        1960      Master       Together  41275.00        1   
2148   7873        1973         PhD       Together  63516.00        1   
2149  10609        1962         PhD        Married  42769.00        0   
2150   6950        1958      Master        Married  49572.00        1   
2151   7233        1972  Graduation        Married  69209.00        0   
2152   9481        1976  Graduation        Married  17148.00        1   
2153   7699        1969  Graduation          Widow  40590.00        1   
2154   2711        1976      Master       Together  30560.00        1   
2155   8933        1983    2n Cycle        Married  32146.00        1   
2156   8726        1964    2n Cycle       Divorced  41713.00        1   
2157   1515        1975      Master       Together  34176.00        0   
2158   5186        1955         PhD        Married  58482.00        0   
2159  10157        1965  Graduation         Single  59686.00        0   
2160   5667        1964         PhD        Married  60896.00        0   
2161    902        1980  Graduation       Together  62994.00        1   
2162    977        1957  Graduation       Together  47320.00        0   
2163   7313        1949    2n Cycle          Widow  74859.00        0   
2164   7600        1992       Basic         Single  15253.00        1   
2165   9706        1974         PhD         Single  31560.00        1   
2166   2429        1954  Graduation       Divorced  72071.00        0   
2167   3520        1990      Master         Single  91172.00        0   
2168  10394        1984  Graduation        Married  90000.00        0   
2169   4548        1981  Graduation         Single  41967.00        1   
2170   9362        1965  Graduation         Single  69263.00        0   
2171   3968        1970    2n Cycle       Divorced  65706.00        0   
2172   4084        1975  Graduation       Together  60934.00        0   
2173   6721        1951  Graduation        Married  71965.00        0   
2174   2829        1955  Graduation        Married  65210.00        0   
2175   1772        1975         PhD        Married  79174.00        0   
2176   1544        1967      Master       Divorced  81380.00        0   
2177    492        1973         PhD           YOLO  48432.00        0   
2178   4377        1971  Graduation        Married  52914.00        0   
2179   4541        1955  Graduation       Together  38946.00        0   
2180   1321        1974  Graduation        Married  26067.00        1   
2181   8858        1972         PhD        Married  44325.00        0   
2182   3266        1964  Graduation        Married  42523.00        0   
2183  11187        1978       Basic         Single  26487.00        1   
2184   8207        1957    2n Cycle        Married  53233.00        0   
2185   9723        1960  Graduation        Married  67716.00        0   
2186   2666        1972      Master        Married  76234.00        0   
2187   5721        1956         PhD        Married  84117.00        0   
2188  10652        1957      Master       Together  65487.00        0   
2189   1646        1972       Basic        Married  25224.00        1   
2190   4418        1983      Master         Single  89616.00        0   
2191   2656        1971  Graduation       Divorced  40851.00        1   
2192   7990        1947  Graduation        Married  27469.00        0   
2193   8722        1957    2n Cycle        Married  82347.00       

<IPython.core.display.Javascript object>

In [12]:
# let's create a copy of the data to avoid any changes to original data
data = df.copy()

<IPython.core.display.Javascript object>

In [13]:
# checking for duplicate values in the data
data.duplicated().sum()

0

<IPython.core.display.Javascript object>

- There are no duplicate values in the data.

In [14]:
# checking the names of the columns in the data
print(data.columns)

Index(['ID', 'Year_Birth', 'Education', 'Marital_Status', 'Income', 'Kidhome',
       'Teenhome', 'Dt_Customer', 'Recency', 'MntWines', 'MntFruits',
       'MntMeatProducts', 'MntFishProducts', 'MntSweetProducts',
       'MntGoldProds', 'NumDealsPurchases', 'NumWebPurchases',
       'NumCatalogPurchases', 'NumStorePurchases', 'NumWebVisitsMonth',
       'AcceptedCmp3', 'AcceptedCmp4', 'AcceptedCmp5', 'AcceptedCmp1',
       'AcceptedCmp2', 'Complain', 'Z_CostContact', 'Z_Revenue', 'Response'],
      dtype='object')


<IPython.core.display.Javascript object>

In [15]:
# checking column datatypes and number of non-null values
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2240 entries, 0 to 2239
Data columns (total 29 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   ID                   2240 non-null   int64  
 1   Year_Birth           2240 non-null   int64  
 2   Education            2240 non-null   object 
 3   Marital_Status       2240 non-null   object 
 4   Income               2216 non-null   float64
 5   Kidhome              2240 non-null   int64  
 6   Teenhome             2240 non-null   int64  
 7   Dt_Customer          2240 non-null   object 
 8   Recency              2240 non-null   int64  
 9   MntWines             2240 non-null   int64  
 10  MntFruits            2240 non-null   int64  
 11  MntMeatProducts      2240 non-null   int64  
 12  MntFishProducts      2240 non-null   int64  
 13  MntSweetProducts     2240 non-null   int64  
 14  MntGoldProds         2240 non-null   int64  
 15  NumDealsPurchases    2240 non-null   i

<IPython.core.display.Javascript object>

* Dependent variable is the Response of a client to a campaign, which if of *int* type.
* `Education`, `Marital_Status`, `Dt_Customer` are of *object* type.
* All other columns are numeric in nature.
* There are missing values in the `income` column.

## Fixing the data types
* `Year_Birth` should be converted in a categorical data type by dividing the column in generations. EX. Millenials, Baby Boomers, GenZ. 

In [16]:
conditions = [
    (df["Year_Birth"] >= 1893) & (df["Year_Birth"] <= 1923),
    (df["Year_Birth"] >= 1924) & (df["Year_Birth"] <= 1945),
    (df["Year_Birth"] >= 1946) & (df["Year_Birth"] <= 1964),
    (df["Year_Birth"] >= 1965) & (df["Year_Birth"] <= 1980),
    (df["Year_Birth"] >= 1981) & (df["Year_Birth"] <= 1996),
    (df["Year_Birth"] >= 1997) & (df["Year_Birth"] <= 2012),
    (df["Year_Birth"] >= 2013) & (df["Year_Birth"] <= 2025),
]

values = [
    "Error",
    "96-100 years old - Silent",
    "59-95 - Boomer",
    "43-58 - GenX",
    "27-57 - Millenialls",
    "11-26 GenZ",
    "0-10 - GenA",
]

<IPython.core.display.Javascript object>

In [17]:
# create new column age_group
data["age_group"] = np.select(conditions, values)

<IPython.core.display.Javascript object>

In [18]:
# Let's drop Year_Birth from our dataset
data.drop(columns=["Year_Birth"], inplace=True)

<IPython.core.display.Javascript object>

In [19]:
data.head(2)

ID   Education Marital_Status   Income  Kidhome  Teenhome Dt_Customer  \
0  5524  Graduation         Single 58138.00        0         0  2012-09-04   
1  2174  Graduation         Single 46344.00        1         1  2014-03-08   

   Recency  MntWines  MntFruits  MntMeatProducts  MntFishProducts  \
0       58       635         88              546              172   
1       38        11          1                6                2   

   MntSweetProducts  MntGoldProds  NumDealsPurchases  NumWebPurchases  \
0                88            88                  3                8   
1                 1             6                  2                1   

   NumCatalogPurchases  NumStorePurchases  NumWebVisitsMonth  AcceptedCmp3  \
0                   10                  4                  7             0   
1                    1                  2                  5             0   

   AcceptedCmp4  AcceptedCmp5  AcceptedCmp1  AcceptedCmp2  Complain  \
0             0             0             0             0         0   
1             0             0             0             0         0   

   Z_CostContact  Z_Revenue  Response       age_group  
0              3         11         1  59-95 - Boomer  
1              3         11         0  59-95 - Boomer

<IPython.core.display.Javascript object>

* `Dt_Customer` should be converted in a categorical data type.

In [58]:
conditions = [
    (df["Dt_Customer"] < "2012-01-01"),
    (df["Dt_Customer"] >= "2012-01-01") & (df["Dt_Customer"] <= "2012-01-31"),
    (df["Dt_Customer"] >= "2013-01-01") & (df["Dt_Customer"] <= "2013-01-31"),
    (df["Dt_Customer"] >= "2014-01-01") & (df["Dt_Customer"] <= "2014-01-31"),
]

values = ["Before 2012", "2012", "2013", "2014"]

<IPython.core.display.Javascript object>

In [59]:
# create new column age_group
data["date_customer"] = np.select(conditions, values)

<IPython.core.display.Javascript object>

In [60]:
data["date_customer"].value_counts()

0       2043
2013     107
2014      90
Name: date_customer, dtype: int64

<IPython.core.display.Javascript object>

In [61]:
data.drop(columns="Dt_Customer", inplace=True)

<IPython.core.display.Javascript object>

In [63]:
data.head(100)

ID   Education Marital_Status   Income  Kidhome  Teenhome  Recency  \
0    5524  Graduation         Single 58138.00        0         0       58   
1    2174  Graduation         Single 46344.00        1         1       38   
2    4141  Graduation       Together 71613.00        0         0       26   
3    6182  Graduation       Together 26646.00        1         0       26   
4    5324         PhD        Married 58293.00        1         0       94   
5    7446      Master       Together 62513.00        0         1       16   
6     965  Graduation       Divorced 55635.00        0         1       34   
7    6177         PhD        Married 33454.00        1         0       32   
8    4855         PhD       Together 30351.00        1         0       19   
9    5899         PhD       Together  5648.00        1         1       68   
10   1994  Graduation        Married      nan        1         0       11   
11    387       Basic        Married  7500.00        0         0       59   
12   2125  Graduation       Divorced 63033.00        0         0       82   
13   8180      Master       Divorced 59354.00        1         1       53   
14   2569  Graduation        Married 17323.00        0         0       38   
15   2114         PhD         Single 82800.00        0         0       23   
16   9736  Graduation        Married 41850.00        1         1       51   
17   4939  Graduation       Together 37760.00        0         0       20   
18   6565      Master        Married 76995.00        0         1       91   
19   2278    2n Cycle         Single 33812.00        1         0       86   
20   9360  Graduation        Married 37040.00        0         0       41   
21   5376  Graduation        Married  2447.00        1         0       42   
22   1993         PhD        Married 58607.00        0         1       63   
23   4047         PhD        Married 65324.00        0         1        0   
24   1409  Graduation       Together 40689.00        0         1       69   
25   7892  Graduation         Single 18589.00        0         0       89   
26   2404  Graduation        Married 53359.00        1         1        4   
27   5255  Graduation         Single      nan        1         0       19   
28   9422  Graduation        Married 38360.00        1         0       26   
29   1966         PhD        Married 84618.00        0         0       96   
30   6864      Master       Divorced 10979.00        0         0       34   
31   3033      Master       Together 38620.00        0         0       56   
32   5710  Graduation       Together 40548.00        0         1       31   
33   7373         PhD       Divorced 46610.00        0         2        8   
34   8755      Master        Married 68657.00        0         0        4   
35  10738      Master         Single 49389.00        1         1       55   
36   4339         PhD        Married 67353.00        0         1       37   
37  10755    2n Cycle        Married 23718.00        1         0       76   
38   8595  Graduation          Widow 42429.00        0         1       99   
39   2968         PhD       Divorced 48948.00        0         0       53   
40   8601  Graduation        Married 80011.00        0         1        3   
41    503      Master        Married 20559.00        1         0       88   
42   8430  Graduation       Together 21994.00        0         1        4   
43   7281         PhD         Single      nan        0         0       80   
44   2139      Master        Married  7500.00        1         0       19   
45   1371  Graduation         Single 79941.00        0         0       72   
46   9909    2n Cycle        Married  7500.00        0         0       24   
47   7286  Graduation       Together 41728.00        1         0       92   
48   7244  Graduation         Single      nan        2         1       96   
49   6566         PhD        Married 72550.00        1         1       39   
50   8614  Graduation          Widow 65486.00        0         1       29   
51 

<IPython.core.display.Javascript object>

* `Education`,`Marital_Status`, `age_group` are of object type, we can change them to categories.

* `AcceptedCmp1`, `AcceptedCmp2`, `AcceptedCmp3`, `AcceptedCmp4`,`AcceptedCmp5`, `Complain`, `Response` are numerical but we can convert them to "category" as well.

* *Coverting "objects" to "category" reduces the data space required to store the dataframe*

In [64]:
data["Education"] = data["Education"].astype("category")
data["Marital_Status"] = data["Marital_Status"].astype("category")
data["age_group"] = data["age_group"].astype("category")
data["AcceptedCmp1"] = data["AcceptedCmp1"].astype("category")
data["AcceptedCmp2"] = data["AcceptedCmp2"].astype("category")
data["AcceptedCmp3"] = data["AcceptedCmp3"].astype("category")
data["AcceptedCmp4"] = data["AcceptedCmp4"].astype("category")
data["AcceptedCmp5"] = data["AcceptedCmp5"].astype("category")
data["Complain"] = data["Complain"].astype("category")
data["Response"] = data["Response"].astype("category")
data["date_customer"] = data["date_customer"].astype("category")

<IPython.core.display.Javascript object>

In [65]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2240 entries, 0 to 2239
Data columns (total 29 columns):
 #   Column               Non-Null Count  Dtype   
---  ------               --------------  -----   
 0   ID                   2240 non-null   int64   
 1   Education            2240 non-null   category
 2   Marital_Status       2240 non-null   category
 3   Income               2216 non-null   float64 
 4   Kidhome              2240 non-null   int64   
 5   Teenhome             2240 non-null   int64   
 6   Recency              2240 non-null   int64   
 7   MntWines             2240 non-null   int64   
 8   MntFruits            2240 non-null   int64   
 9   MntMeatProducts      2240 non-null   int64   
 10  MntFishProducts      2240 non-null   int64   
 11  MntSweetProducts     2240 non-null   int64   
 12  MntGoldProds         2240 non-null   int64   
 13  NumDealsPurchases    2240 non-null   int64   
 14  NumWebPurchases      2240 non-null   int64   
 15  NumCatalogPurchases  

<IPython.core.display.Javascript object>

* *we can see that the memory usage has decreased from 507.6+KB to 390.9+KB*

**Let's check for missing values in the data.**

# checking for missing values in the data

In [66]:
data.isnull().sum()

ID                      0
Education               0
Marital_Status          0
Income                 24
Kidhome                 0
Teenhome                0
Recency                 0
MntWines                0
MntFruits               0
MntMeatProducts         0
MntFishProducts         0
MntSweetProducts        0
MntGoldProds            0
NumDealsPurchases       0
NumWebPurchases         0
NumCatalogPurchases     0
NumStorePurchases       0
NumWebVisitsMonth       0
AcceptedCmp3            0
AcceptedCmp4            0
AcceptedCmp5            0
AcceptedCmp1            0
AcceptedCmp2            0
Complain                0
Z_CostContact           0
Z_Revenue               0
Response                0
age_group               0
date_customer           0
dtype: int64

<IPython.core.display.Javascript object>

* `income` column has 24 missing values. It represents 1% of total data in the column.
* No other column has missing values.

In [67]:
# Let's look at the statistical summary of the data
data.describe().T

count     mean      std     min      25%      50%  \
ID                  2240.00  5592.16  3246.66    0.00  2828.25  5458.50   
Income              2216.00 52247.25 25173.08 1730.00 35303.00 51381.50   
Kidhome             2240.00     0.44     0.54    0.00     0.00     0.00   
Teenhome            2240.00     0.51     0.54    0.00     0.00     0.00   
Recency             2240.00    49.11    28.96    0.00    24.00    49.00   
MntWines            2240.00   303.94   336.60    0.00    23.75   173.50   
MntFruits           2240.00    26.30    39.77    0.00     1.00     8.00   
MntMeatProducts     2240.00   166.95   225.72    0.00    16.00    67.00   
MntFishProducts     2240.00    37.53    54.63    0.00     3.00    12.00   
MntSweetProducts    2240.00    27.06    41.28    0.00     1.00     8.00   
MntGoldProds        2240.00    44.02    52.17    0.00     9.00    24.00   
NumDealsPurchases   2240.00     2.33     1.93    0.00     1.00     2.00   
NumWebPurchases     2240.00     4.08     2.78    0.00     2.00     4.00   
NumCatalogPurchases 2240.00     2.66     2.92    0.00     0.00     2.00   
NumStorePurchases   2240.00     5.79     3.25    0.00     3.00     5.00   
NumWebVisitsMonth   2240.00     5.32     2.43    0.00     3.00     6.00   
Z_CostContact       2240.00     3.00     0.00    3.00     3.00     3.00   
Z_Revenue           2240.00    11.00     0.00   11.00    11.00    11.00   

                         75%       max  
ID                   8427.75  11191.00  
Income              68522.00 666666.00  
Kidhome                 1.00      2.00  
Teenhome                1.00      2.00  
Recency                74.00     99.00  
MntWines              504.25   1493.00  
MntFruits              33.00    199.00  
MntMeatProducts       232.00   1725.00  
MntFishProducts        50.00    259.00  
MntSweetProducts       33.00    263.00  
MntGoldProds           56.00    362.00  
NumDealsPurchases       3.00     15.00  
NumWebPurchases         6.00     27.00  
NumCatalogPurchases     4.00     28.00  
NumStorePurchases       8.00     13.00  
NumWebVisitsMonth       7.00     20.00  
Z_CostContact           3.00      3.00  
Z_Revenue              11.00     11.00

<IPython.core.display.Javascript object>

* We have 2,240 clients

* We can see that the `Response` is a binary of 1 and 0.  14% of our clients engaged in the campaign.

* 50% of our clients were born before 1970.

* 50% of our clients' `income` sits below $51,000.  However, `income` has a wide range from 1730 to 666666

* The product category that our clients invest the greatest dollar amoount is wines.  

* At least 50% of our clients have spent 173 or less in wines.  

* The dollar amount each client spends in wine has a wide range 0 to  1493. 
    - This might be due to outliers in our data

* The second best product category in dollar amoount is Meat.  50% of our clients have spent 67 or less in meat.

* At least 50% of our clients have 0 kids at home

* At least 50% of our clients have:

    * 6 web visits a month

    * 5 in-store purchases

    * 2 catalog purchases

    * 4 web purchases

    * 2 num deals purchases

    * 0 teens at home


**Let's look at the non-numeric columns.**

In [68]:
# filtering non-numeric columns
cat_columns = data.select_dtypes(exclude=np.number).columns
cat_columns

Index(['Education', 'Marital_Status', 'AcceptedCmp3', 'AcceptedCmp4',
       'AcceptedCmp5', 'AcceptedCmp1', 'AcceptedCmp2', 'Complain', 'Response',
       'age_group', 'date_customer'],
      dtype='object')

<IPython.core.display.Javascript object>

In [69]:
# printing the number of occurrences of each unique value in each categorical column
for column in cat_columns:

    print(data[column].value_counts() / len(data) * 100)
    print("-" * 50)

Graduation   50.31
PhD          21.70
Master       16.52
2n Cycle      9.06
Basic         2.41
Name: Education, dtype: float64
--------------------------------------------------
Married    38.57
Together   25.89
Single     21.43
Divorced   10.36
Widow       3.44
Alone       0.13
YOLO        0.09
Absurd      0.09
Name: Marital_Status, dtype: float64
--------------------------------------------------
0   92.72
1    7.28
Name: AcceptedCmp3, dtype: float64
--------------------------------------------------
0   92.54
1    7.46
Name: AcceptedCmp4, dtype: float64
--------------------------------------------------
0   92.72
1    7.28
Name: AcceptedCmp5, dtype: float64
--------------------------------------------------
0   93.57
1    6.43
Name: AcceptedCmp1, dtype: float64
--------------------------------------------------
0   98.66
1    1.34
Name: AcceptedCmp2, dtype: float64
--------------------------------------------------
0   99.06
1    0.94
Name: Complain, dtype: float64
-----------------

<IPython.core.display.Javascript object>

- Master, PHD and Graduation represent more than 71% of our clients.
- Top three marital status groups: 38% of our clients are married, 25% are together, and 21% are single.

### We will drop the missing values in the dataset.

In [70]:
data.dropna(inplace=True)
data.shape

(2216, 29)

<IPython.core.display.Javascript object>

## Let's visualize the data

### Univariate Analysis

In [71]:
# function to plot a boxplot and a histogram along the same scale.


def histogram_boxplot(data, feature, figsize=(12, 7), kde=False, bins=None):
    """
    Boxplot and histogram combined

    data: dataframe
    feature: dataframe column
    figsize: size of figure (default (12,7))
    kde: whether to the show density curve (default False)
    bins: number of bins for histogram (default None)
    """
    f2, (ax_box2, ax_hist2) = plt.subplots(
        nrows=2,  # Number of rows of the subplot grid= 2
        sharex=True,  # x-axis will be shared among all subplots
        gridspec_kw={"height_ratios": (0.25, 0.75)},
        figsize=figsize,
    )  # creating the 2 subplots
    sns.boxplot(
        data=data, x=feature, ax=ax_box2, showmeans=True, color="violet"
    )  # boxplot will be created and a star will indicate the mean value of the column
    sns.histplot(
        data=data, x=feature, kde=kde, ax=ax_hist2, bins=bins, palette="winter"
    ) if bins else sns.histplot(
        data=data, x=feature, kde=kde, ax=ax_hist2
    )  # For histogram
    ax_hist2.axvline(
        data[feature].mean(), color="green", linestyle="--"
    )  # Add mean to the histogram
    ax_hist2.axvline(
        data[feature].median(), color="black", linestyle="-"
    )  # Add median to the histogram

<IPython.core.display.Javascript object>

### `rating`

In [ ]:
a = data.columns

In [ ]:
for i in a
    histogram_boxplot(data, i)

### <a id='link1'>Summary of EDA</a>

**Data Description:**

- The target variable (`rating`) is of *float* type.
- `title`, `description`, `mediaType`, `sznOfRelease`, and `studio_primary` are of *object* type.
- `ongoing` column is of *bool* type.
- All other columns are numeric in nature.
- The `title` and `description` columns are dropped for modeling as they are highly textual in nature.
- There are no duplicate values in the data.
- There are missing values in the data. The rows with missing data have been dropped.


**Observations from EDA:**

- `rating`: The anime ratings are close to normally distributed, with a mean rating of 2.74. The rating increases with an increase in the number of people who have watched or want to watch the anime.
- `eps`: The distribution is heavily right-skewed as there are many anime movies in the data (at least 50%), and they are considered to be of only one episode as per data description. The number of episodes increases as the anime runs for more years.
- `duration`: The distribution is right-skewed with a median anime runtime of less than 10 minutes. With the increase in rating, the duration increases.
- `years_running`: The distribution is heavily right-skewed, and at least 75% of the anime have run for less than 1 year.
- `watched`: The distribution is heavily right-skewed, and most of the anime have less than 500 viewers. This attribute is highly correlated with the `wantWatch` and `votes` attributes.
- `watching`: The distribution is heavily right-skewed and highly correlated with the `dropped` attribute.
- `wantWatch`: The distribution is heavily right-skewed with a median value of 132 potential watchers.
- `dropped`: The distribution is heavily right-skewed with a drop of 25 viewers on average.
- `votes`: The distribution is heavily right-skewed, and few shows have more than 5000 votes.
- `mediaType`: 23% of the anime are published for TV, 17% as music videos, and 14% as web series. Anime available as TV series, web series, or music videos have a lower rating in general
- `ongoing`: Less than 1% of the anime in the data are ongoing.
- `sznOfRelease`: The season of release is missing for nearly 90% of the anime in the data, and is spread out almost evenly across all seasons when available. Anime ratings have a similar distribution across all the seasons of release.
- `studio_primary`: Nearly 40% of the anime in the data are produced by studios not listed in the data. Toei Animation is the most common studio among the available studio names. In general, the ratings are low for anime produced by DLE studios and studios other than the ones listed in the data.
- `studios_colab`: More than 95% of the anime in the data do not involve collaboration between studios.
- `contentWarn`: Less than 10% of the anime in the data have an associated content warning.
- `tag_<tag/genre>`: There are 1747 anime that are based on manga, 1920 of the Comedy genre, 1238 of the Action genre, 1079 anime of the Romance genre, and more.

## Analysis

## Conclusions

## Recommendations